### ISSUE TITLE
Convert to es6 modules

### ISSUE THREAD

**Issue by rkusa** [rkusa/pdfjs#87]
Before leaving alpha, the library should be converted to using es6 modules.

**Comment by rkusa** [comment]
Anyone aware of a tool to automatically convert a library to es6 modules?

**Comment by rkusa** [comment]
While I did an initial conversion, I am not yet sure how to handle the fact that pdfjs should work for both server and client. Since Node.js requires the file extension of `.mjs`

**Comment by jpbourgeon** [comment]
Hello
have a look here 
https://github.com/nolanlawson/cjs-to-es6

Other ideas on stackoverflow
https://stackoverflow.com/questions/43106130/convert-existing-legacy-library-to-es6-module

**Comment by rkusa** [comment]
Hi @jpbourgeon, thanks for the links. I have already converted the library in a local branch, but was not able to find a good es6 modules solution for libraries that target both server (Node.js) and client 😕

**Comment by jpbourgeon** [comment]
Would it help to split the library into two components and take advantage of npmjs namespaces?Something like @pdfjs/server and @pdfjs/browser

**Comment by rkusa** [comment]
I guess there are the following options.

**1) Write client-side ES modules (.js)**

- ✅  does work with babel
- ✅ named imports work (e.g. `import {Document} from "pdfjs"`
- ❌ does not work with Node.js ES modules (since .mjs extension required)
- ❌ does not work with `require` 
- ➡️ the code has to be transpiled back to common-js and additionally included in the module

**2) Write server-side ES modules (.mjs)**

- ✅ does work with babel
- ✅ named imports work (e.g. `import {Document} from "pdfjs"`
- ✅ does work with Node.js ES modules
- ❌ does not work with `require`
- ➡️ the code still has to be transpiled back to CJS modules to get `require` to work

**3) Keep CJS modules**

- ✅ does work with babel
- ❌ named imports do not work (e.g. `import {Document} from "pdfjs"`)
- ✅ can be loaded from Node.js ES modules
- ✅ does work with `require`
- ➡️ only drawback seems to be that we have to write `import * as pdf from "pdfjs"` when loading `pdfjs` from ES modules (instead of being able to import named exports `import {Document} from "pdfjs"`)

In Summary, either transpile the code for Node.js usage or live with `import {Document} from "pdfjs"`. TBH, I really dislike both options 😕

**Comment by rkusa** [comment]
The conversion happens here https://github.com/rkusa/pdfjs/tree/es6-modules
Though, the branch is unfortunately not green yet

**Comment by AllNamesRTaken** [comment]
Would it be of interest to move to TS instead of ES6?

**Comment by amotl** [comment]
> Would it be of interest to move to TS instead of ES6?

I've recently had the pleasure to work on a Typescript codebase [1] using WebStorm [2] and I want to report that the overall experience is absolutely amazing. Finally, I feel like I can get similar productive with Javascript as I am with Python. ES6 did not give me that feeling yet.

[1] https://github.com/hiveeyes/grafana-worldmap-panel
[2] https://www.jetbrains.com/webstorm/

**Comment by AllNamesRTaken** [comment]
My pet project is https://github.com/AllNamesRTaken/GoodCore outside workhours and i could help out is needed.

**Comment by amotl** [comment]
@AllNamesRTaken: Sounds interesting, thanks for letting us know. Star-given ;].

**Comment by rkusa** [comment]
> Would it be of interest to move to TS instead of ES6?

I am, generally speaking, not opposed to TS.

> I've recently had the pleasure to work on a TypeScript codebase [1] using WebStorm [2] and I want to report that the overall experience is absolutely amazing and totally washed away any hesitations I had regarding TypeScript.

I can confirm that. In my day job, I also work more with TS than JS and the whole TS dev experience (editor integration, less errors when refactoring, ...) is superior than JS ...

---

What currently holds me off from converting the codebase to TS:

- I always disliked Coffeescript projects (example works for all other languages that compile to JS), because of the higher entry barrier for submitting a contribution - I am a bit concerned that it might be similar for people when it comes to TS ... though, I think the advantages of TS a greater than my concerns here
- The effort required to convert it
- I am not sure if I'd rather rewrite the lib in something that compiles to wasm (something that would be cooler for my personal use-case of the lib)
